### DISCUSSION TITLE
using cuda or gpu for xgboost on ./src/Train_Models

### DISCUSSION THREAD

**Discussion by fuzmik** [kyleskom/NBA-Machine-Learning-Sports-Betting#434] [Ideas]
I want to use param 'device:cuda' for faster training. can someone check my edited code.
`import sqlite3

import numpy as np
import pandas as pd
import xgboost as xgb
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from tqdm import tqdm

dataset = "dataset_2012-24"
con = sqlite3.connect("../../Data/dataset.sqlite")
data = pd.read_sql_query(
    f"select * from \"{dataset}\"", con, index_col="index")
con.close()

margin = data['Home-Team-Win']
data.drop(['Score', 'Home-Team-Win', 'TEAM_NAME', 'Date', 'TEAM_NAME.1', 'Date.1', 'OU-Cover', 'OU'],
          axis=1, inplace=True)

data = data.values

data = data.astype(float)
highest_acc = 0
best_model = None

for x in tqdm(range(300)):
    x_train, x_test, y_train, y_test = train_test_split(
        data, margin, test_size=.1)

    train = xgb.DMatrix(x_train, label=y_train)
    test = xgb.DMatrix(x_test, label=y_test)

    param = {
        'max_depth': 3,
        'eta': 0.01,
        'objective': 'multi:softprob',
        'num_class': 2,
        'device': 'cuda',  # Use GPU for prediction
        'tree_method': 'hist'  # Use GPU for training
    }
    paramtest = {
        'device': 'cuda',  # Use GPU for prediction
        'tree_method': 'hist'  # Use GPU for training
    }
    epochs = 750

    model = xgb.train(param, train, epochs)
    predictions = model.predict(test, paramtest)
    y = []

    for z in predictions:
        y.append(np.argmax(z))

    acc = round(accuracy_score(y_test, y) * 100, 1)
    print(f"{acc}%")
    if acc > highest_acc:
        highest_acc = acc
        best_model = model

if best_model is not None:
    best_model.save_model(
        f'../../Models/XGBoost_Models/XGBoost_{highest_acc}%_ML-4.json')`

my tensorflow does detect the gpu.
my coding is very limited understanding. hope someone can check. the code does run. i just wonder if the model result is good.

**Comment by kyleskom** [comment]
I use a Mac with apple silicon so sadly I cant provide any guidance here. Hopefully someone else can!

**Reply by fuzmik** [reply]
seems like better to use CPU. my ollama saying something about blah blah predictions better using CPU. Another question. if I were to use ollama using retrieval augmented generator (RAG). what  documents or website links would suggest or keywords. I'm trying to place the ollama response on to flask. I've tried giving it scoresandodds.com espn.com NBA.com. and maybe the SQL DB.
